### PR TITLE
Make latest.jpg redirect to last meme generated (WIP)

### PR DESCRIPTION
**Pull request by joshfriend**
TODO:

* [ ] Use in-memory (FakeRedis?) for development/tests
* [ ] Write unit tests

### CODE REVIEW

File: memegen/services/image.py [hunk 1]
```diff
@@ -22,6 +23,13 @@ def find_template(self, key):
     def create_image(self, template, text, style=None):
         image = Image(template, text, style=style)
 
+        data = {
+            'path': text.path,
+            'template': template.key,
+            'style': style,
+        }
+        redis.hmset('latest', data)
```
Comment (jacebrowning) [inline memegen/services/image.py]: Is there a reason we write to redis in the service, but read from redis in the store?

File: memegen/stores/image.py [hunk 2]
```diff
@@ -1,16 +1,44 @@
...
         self.root = root
 
     @property
     def latest(self):
-        return os.path.join(self.root, self.LATEST)
+        data = redis.hgetall('latest')
+        if not data:
+            raise FileNotFoundError()
+
+        data = decode_bytstring_dict(data)
```
Comment (jacebrowning) [inline memegen/stores/image.py]: Perhaps we don't even need a function?

`data = {k.decode(): v.decode() for k,v in data.items()}`

File: memegen/settings.py [hunk 3]
```diff
@@ -13,6 +13,8 @@ class Config:
 
     GOOGLE_ANALYTICS_TID = os.getenv('GOOGLE_ANALYTICS_TID')
 
+    REDIS_URL = os.getenv('REDIS_URL', 'redis://localhost:6379/0')
```
Comment (jacebrowning) [inline memegen/settings.py]: Where is this going to be used?

**Comment by jacebrowning** [comment]
This is looking really good! In what cases should Redis be running locally? Does it have a way to fallback to an in-memory store for local development?

File: memegen/services/image.py (hunk 1 above)
Comment (joshfriend) [inline memegen/services/image.py]: Because the store gets skipped if the image already exists on disk

File: memegen/settings.py (hunk 3 above)
Comment (joshfriend) [inline memegen/settings.py]: It gets used by the Flask-Redis extension.

**Comment by joshfriend** [comment]
I've still got some things to work out (see list in OP), but I'm going to try to find a way to use a fake in-memory Redis for tests, and when the user doesn't set `REDIS_URL` in their `.env`. It would work best for development to use a real Redis instance since an in-memory fake would not be shared across Gunicorn processes.

File: memegen/services/image.py (hunk 1 above)
Comment (jacebrowning) [inline memegen/services/image.py]: I think I'd like to refactor this in `develop` so that it becomes the responsibility of the store.

Comment (joshfriend) [inline memegen/services/image.py]: Thats probably a good idea. I just thought of a way to put all the redis stuff in the store, but I have to play with it some more...

Comment (jacebrowning) [inline memegen/services/image.py]: Here's the refactor: https://github.com/jacebrowning/memegen/pull/157

If you wouldn't mind, please give that a quick look before I merge. You'll need to rebase after that.

Comment (jacebrowning) [inline memegen/services/image.py]: Merged.

**Comment by jacebrowning** [comment]
Do you still plan to work on this? I'm happy to take it over if not.

**Comment by joshfriend** [comment]
I don't expect to have enough time to finish this in the near future :(

Feel free to continue if you feel it's worthwhile...

**Comment by jacebrowning** [comment]
No worries! I do think it's valuable so I'll pick this up.

**Pull request closed without merging**